File: android/src/main/java/com/easemob/im_flutter_sdk/PushManagerWrapper.java
package com.easemob.im_flutter_sdk;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMPushConfigs;
import com.hyphenate.chat.EMPushManager.DisplayStyle;
import com.hyphenate.chat.EMSilentModeParam;
import com.hyphenate.chat.EMSilentModeResult;
import com.hyphenate.exceptions.HyphenateException;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;


public class PushManagerWrapper extends Wrapper implements MethodCallHandler {

    PushManagerWrapper(FlutterPlugin.FlutterPluginBinding flutterPluginBinding, String channelName) {
        super(flutterPluginBinding, channelName);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        JSONObject param = (JSONObject)call.arguments;
        try {
            if (MethodKey.getImPushConfig.equals(call.method)) {
                getImPushConfig(param, call.method, result);
            }
            else if(MethodKey.getImPushConfigFromServer.equals(call.method)){
                getImPushConfigFromServer(param, call.method, result);
            }
            else if(MethodKey.updatePushNickname.equals(call.method)){
                updatePushNickname(param, call.method, result);
            }
            else if(MethodKey.updateImPushStyle.equals(call.method)){
                updateImPushStyle(param, call.method, result);
            }
            else if(MethodKey.updateHMSPushToken.equals(call.method)){
                updateHMSPushToken(param, call.method, result);
            }
            else if(MethodKey.updateFCMPushToken.equals(call.method)){
                updateFCMPushToken(param, call.method, result);
            }
            else if (MethodKey.reportPushAction.equals(call.method)) {
                reportPushAction(param, call.method, result);
            }
            else if (MethodKey.setConversationSilentMode.equals(call.method)) {
                setConversationSilentMode(param, call.method, result);
            }
            else if (MethodKey.removeConversationSilentMode.equals(call.method)) {
                removeConversationSilentMode(param, call.method, result);
            }
            else if (MethodKey.fetchConversationSilentMode.equals(call.method)) {
                fetchConversationSilentMode(param, call.method, result);
            }
            else if (MethodKey.setSilentModeForAll.equals(call.method)) {
                setSilentModeForAll(param, call.method, result);
            }
            else if (MethodKey.fetchSilentModeForAll.equals(call.method)) {
                fetchSilentModeForAll(param, call.method, result);
            }
            else if (MethodKey.fetchSilentModeForConversations.equals(call.method)) {
                fetchSilentModeForConversations(param, call.method, result);
            }
            else if (MethodKey.setPreferredNotificationLanguage.equals(call.method)) {
                setPreferredNotificationLanguage(param, call.method, result);
            }
            else if (MethodKey.fetchPreferredNotificationLanguage.equals(call.method)) {
                fetchPreferredNotificationLanguage(param, call.method, result);
            }
            else if (MethodKey.getPushTemplate.equals(call.method)) {
                getPushTemplate(param, call.method, result);
            }
            else if (MethodKey.setPushTemplate.equals(call.method)) {
                setPushTemplate(param, call.method, result);
            }
            // 481
            else if (MethodKey.syncSilentModels.equals(call.method)) {
                syncSilentModels(param, call.method, result);
            }
            else if (MethodKey.bindDeviceToken.equals(call.method)) {
                bindDeviceToken(param, call.method, result);
            }
            else {
                super.onMethodCall(call, result);
            }
        }catch (JSONException e) {

        }
    }

    private void getImPushConfig(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(()->{
            EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigs();
            onSuccess(result, channelName, PushConfigsHelper.toJson(configs));
        });

    }

    private void getImPushConfigFromServer(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(()->{
            try {
                EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigsFromServer();
                onSuccess(result, channelName, PushConfigsHelper.toJson(configs));
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void updatePushNickname(JSONObject params, String channelName,  Result result) throws JSONException {
        String nickname = params.getString("nickname");

        asyncRunnable(()->{
            try {
                EMClient.getInstance().pushManager().updatePushNickname(nickname);
                onSuccess(result, channelName, nickname);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }


    private void updateImPushStyle(JSONObject params, String channelName,  Result result) throws JSONException {
        DisplayStyle style = params.getInt("pushStyle") == 0 ? DisplayStyle.SimpleBanner : DisplayStyle.MessageSummary;
        EMClient.getInstance().pushManager().asyncUpdatePushDisplayStyle(style, new EMWrapperCallBack(result, channelName, true));
    }


    private void updateHMSPushToken(JSONObject params, String channelName,  Result result) throws JSONException {
        String token = params.getString("token");
        asyncRunnable(()->{
            EMClient.getInstance().sendHMSPushTokenToServer(token);
            onSuccess(result, channelName, token);
        });
    }

    private void updateFCMPushToken(JSONObject params, String channelName,  Result result) throws JSONException {
        String token = params.getString("token");
        String fcmKey = EMClient.getInstance().getOptions().getPushConfig().getFcmSenderId();
        EMClient.getInstance().pushManager().bindDeviceToken(fcmKey, token, new EMWrapperCallBack(result, channelName, null));
    }

    private void reportPushAction(JSONObject params, String channelName, Result result) throws JSONException {

    }

    private void setConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EnumTools.conversationTypeFromInt(params.getInt("conversationType"));
        EMSilentModeParam param = SilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForConversation(conversationId, type, param, new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(null);
            }
        });
    }
    private void removeConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EnumTools.conversationTypeFromInt(params.getInt("conversationType"));
        EMClient.getInstance().pushManager().clearRemindTypeForConversation(conversationId, type, new EMWrapperCallBack(result, channelName, null));
    }

    private void fetchConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EnumTools.conversationTypeFromInt(params.getInt("conversationType"));
        EMClient.getInstance().pushManager().getSilentModeForConversation(conversationId, type, new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(SilentModeResultHelper.toJson(object));
            }
        });
    }

    private void setSilentModeForAll(JSONObject params, String channelName, Result result) throws JSONException {
        EMSilentModeParam param =  SilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForAll(param ,new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(null);
            }
        });
    }

    private void fetchSilentModeForAll(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getSilentModeForAll(new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(SilentModeResultHelper.toJson(object));
            }
        });
    }
    private void fetchSilentModeForConversations(JSONObject params, String channelName, Result result) throws JSONException {
        Iterator iterator = params.keys();
        ArrayList<EMConversation> list = new ArrayList<>();
        while (iterator.hasNext()) {
            String conversationId = (String)iterator.next();
            EMConversation.EMConversationType type = EnumTools.conversationTypeFromInt(params.getInt(conversationId));
            EMConversation conversation = EMClient.getInstance().chatManager().getConversation(conversationId, type, true);
            list.add(conversation);
        }

        EMClient.getInstance().pushManager().getSilentModeForConversations(list, new EMValueWrapperCallBack<Map<String, EMSilentModeResult>>(result, channelName) {
            @Override
            public void onSuccess(Map<String, EMSilentModeResult> object) {
                Map<String ,Map> result = new HashMap<>();
                for (Map.Entry<String, EMSilentModeResult>entry: object.entrySet()) {
                    result.put(entry.getKey(), SilentModeResultHelper.toJson(entry.getValue()));
                }
                super.updateObject(result);
            }
        });

    }

    private void setPreferredNotificationLanguage(JSONObject params, String channelName, Result result) throws JSONException {
        String code = params.getString("code");
        EMClient.getInstance().pushManager().setPreferredNotificationLanguage(code, new EMWrapperCallBack(result, channelName, null));
    }

    private void fetchPreferredNotificationLanguage(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getPreferredNotificationLanguage(new EMValueWrapperCallBack<String>(result, channelName){
            @Override
            public void onSuccess(String object) {
                super.onSuccess(object);
            }
        });
    }

    private void setPushTemplate(JSONObject params, String channelName, Result result) throws JSONException {
        String pushTemplateName = params.getString("pushTemplateName");
        EMClient.getInstance().pushManager().setPushTemplate(pushTemplateName, new EMWrapperCallBack(result, channelName, null));
    }

    private void getPushTemplate(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getPushTemplate(new EMValueWrapperCallBack<String>(result, channelName) {
            @Override
            public void onSuccess(String object) {
                super.onSuccess(object);
            }
        });
    }

    // 481
    private void syncSilentModels(JSONObject params, String channelName, Result result) {
        EMClient.getInstance().pushManager().syncSilentModeConversationsFromServer(new EMWrapperCallBack(result, channelName, null));
    }

    private void bindDeviceToken(JSONObject params, String channelName, Result result)  throws JSONException {
        String notifierName = params.getString("notifierName");
        String deviceToken = params.getString("deviceToken");
        EMClient.getInstance().pushManager().bindDeviceToken(notifierName, deviceToken, new EMWrapperCallBack(result, channelName, null));
    }

}

File: im_flutter_sdk_ios/ios/Classes/PushManagerWrapper.m
//
//  EMPushManagerWrapper.m
//  im_flutter_sdk
//
//  Created by 东海 on 2020/5/7.
//

#import "PushManagerWrapper.h"
#import "MethodKeys.h"
#import "PushOptionsHelper.h"
#import "ConversationHelper.h"
#import "GroupHelper.h"
#import "SilentModeParamHelper.h"
#import "SilentModeResultHelper.h"
#import "EnumTools.h"

@implementation PushManagerWrapper

- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                               registrar:registrar]) {
    }
    return self;
}

#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call
                  result:(FlutterResult)result {
    if ([ChatGetImPushConfig isEqualToString:call.method]) {
        [self getImPushConfig:call.arguments
                  channelName:call.method
                       result:result];
    } else if ([ChatGetImPushConfigFromServer isEqualToString:call.method]) {
        [self getImPushConfigFromServer:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatUpdatePushNickname isEqualToString:call.method]) {
        [self updatePushNickname:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([ChatUpdateImPushStyle isEqualToString:call.method]) {
        [self updateImPushStyle:call.arguments
                    channelName:call.method
                         result:result];
    } else if ([ChatBindDeviceToken isEqualToString:call.method] || [ChatUpdateFCMPushToken isEqualToString:call.method]) {
        [self  bindAPNsDeviceToken:call.arguments
                       channelName:call.method
                            result:result];
    } else if ([ChatUpdateFCMPushToken isEqualToString:call.method]) {
        [self  bindFCMToken:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatReportPushAction isEqualToString:call.method]){
        [self reportPushAction:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatSetConversationSilentMode isEqualToString:call.method]){
        [self setConversationSilentMode:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatRemoveConversationSilentMode isEqualToString:call.method]){
        [self removeConversationSilentMode:call.arguments
                               channelName:call.method
                                    result:result];
    } else if ([ChatFetchConversationSilentMode isEqualToString:call.method]){
        [self fetchConversationSilentMode:call.arguments
                              channelName:call.method
                                   result:result];
    } else if ([ChatSetSilentModeForAll isEqualToString:call.method]){
        [self setSilentModeForAll:call.arguments
                      channelName:call.method
                           result:result];
    } else if ([ChatFetchSilentModeForAll isEqualToString:call.method]){
        [self fetchSilentModeForAll:call.arguments
                        channelName:call.method
                             result:result];
    } else if ([ChatFetchSilentModeForConversations isEqualToString:call.method]){
        [self fetchSilentModeForConversations:call.arguments
                                  channelName:call.method
                                       result:result];
    } else if ([ChatSetPreferredNotificationLanguage isEqualToString:call.method]){
        [self setPreferredNotificationLanguage:call.arguments
                                   channelName:call.method
                                        result:result];
    } else if ([ChatFetchPreferredNotificationLanguage isEqualToString:call.method]){
        [self fetchPreferredNotificationLanguage:call.arguments
                                     channelName:call.method
                                          result:result];
    } 
    else if ([ChatSetPushTemplate isEqualToString:call.method]) {
        [self setPushTemplate:call.arguments channelName:call.method result:result];
    } 
    else if ([ChatGetPushTemplate isEqualToString:call.method]) {
        [self getPushTemplate:call.arguments channelName:call.method result:result];
    }
    // 481
    else if([ChatSyncSilentModels isEqualToString:call.method]) {
        [self syncConversationsSilentMode:call.arguments channelName:call.method result:result];
    }
    else if([bindDeviceToken isEqualToString:call.method]) {
        [self bindDeviceToken:call.arguments channelName:call.method result:result];
    }
    else{
        [super handleMethodCall:call result:result];
    }
}

- (void)getImPushConfig:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    EMPushOptions *options = EMClient.sharedClient.pushManager.pushOptions;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:[options toJson]];
}

- (void)getImPushConfigFromServer:(NSDictionary *)param
                      channelName:(NSString *)aChannelName
                           result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPushNotificationOptionsFromServerWithCompletion:^(EMPushOptions *aOptions, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aOptions toJson]];
        
    }];
}

- (void)updatePushNickname:(NSDictionary *)param
               channelName:(NSString *)aChannelName
                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *nickname = param[@"nickname"];
    [EMClient.sharedClient.pushManager updatePushDisplayName:nickname
                                                  completion:^(NSString * _Nonnull aDisplayName, EMError * _Nonnull aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aDisplayName];
    }];
    
}


- (void)updateImPushStyle:(NSDictionary *)param
              channelName:(NSString *)aChannelName
                   result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    
    EMPushDisplayStyle pushStyle = [param[@"pushStyle"] intValue];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *aError = [EMClient.sharedClient.pushManager updatePushDisplayStyle:pushStyle];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:@(!aError)];
        });
    });
}

- (void)bindAPNsDeviceToken:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *deviceToken = param[@"token"];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *error = [EMClient.sharedClient bindDeviceToken:deviceToken];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:error
                               object:nil];
        });
    });
}

- (void)bindFCMToken:(NSDictionary *)param
         channelName:(NSString *)aChannelName
              result:(FlutterResult)result {
 __weak typeof(self) weakSelf = self;
 NSString *deviceToken = param[@"token"];
 dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
     [EMClient.sharedClient bindFCMToken:deviceToken completion:^(EMError * _Nullable aError) {
         [weakSelf wrapperCallBack:result
                       channelName:aChannelName
                             error:aError
                            object:nil];
     }];
 });
}


- (void)reportPushAction:(NSDictionary *)param
             channelName:(NSString *)aChannelName
                  result:(FlutterResult)result {
    
}

- (void)setConversationSilentMode:(NSDictionary *)param
                      channelName:(NSString *)aChannelName
                           result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EnumTools conversationTypeFromInt:[param[@"conversationType"] intValue]];
    EMSilentModeParam *silmentParam = [EMSilentModeParam fromJson:param[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                             params:silmentParam
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)removeConversationSilentMode:(NSDictionary *)param
                         channelName:(NSString *)aChannelName
                              result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EnumTools conversationTypeFromInt:[param[@"conversationType"] intValue]];
    [EMClient.sharedClient.pushManager clearRemindTypeForConversation:conversaionId
                                                     conversationType:type
                                                           completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}
- (void)fetchConversationSilentMode:(NSDictionary *)param
                        channelName:(NSString *)aChannelName
                             result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EnumTools conversationTypeFromInt:[param[@"conversationType"] intValue]];
    [EMClient.sharedClient.pushManager getSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aResult toJson]];
    }];
}

- (void)setSilentModeForAll:(NSDictionary *)param
                channelName:(NSString *)aChannelName
                     result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    EMSilentModeParam *silmentParam = [EMSilentModeParam fromJson:param[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForAll:silmentParam
                                                completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)fetchSilentModeForAll:(NSDictionary *)param
                  channelName:(NSString *)aChannelName
                       result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getSilentModeForAllWithCompletion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aResult toJson]];
    }];
}

- (void)fetchSilentModeForConversations:(NSDictionary *)param
                            channelName:(NSString *)aChannelName
                                 result:(FlutterResult)result {
    NSMutableArray *conversations = [NSMutableArray array];
    for (NSString *conversaitonId in param.allKeys) {
        EMConversationType type = [EnumTools conversationTypeFromInt:[param[conversaitonId] intValue]];
        EMConversation *conversation = [EMClient.sharedClient.chatManager getConversation:conversaitonId type:type createIfNotExist:YES];
        [conversations addObject:conversation];
    }
    
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getSilentModeForConversations:conversations
                                                          completion:^(NSDictionary<NSString *,EMSilentModeResult *> * _Nullable aResult, EMError * _Nullable aError)
     {
        
        NSMutableDictionary *tmpDict = [NSMutableDictionary dictionary];
        for (NSString *conversationId in aResult.allKeys) {
            tmpDict[conversationId] = [aResult[conversationId] toJson];
        }
        
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:tmpDict];
    }];
}

- (void)setPreferredNotificationLanguage:(NSDictionary *)param
                             channelName:(NSString *)aChannelName
                                  result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *code = param[@"code"];
    [EMClient.sharedClient.pushManager setPreferredNotificationLanguage:code completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)fetchPreferredNotificationLanguage:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPreferredNotificationLanguageCompletion:^(NSString * _Nullable aLaguangeCode, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aLaguangeCode];
    }];
}

- (void)setPushTemplate:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *pushTemplateName = param[@"pushTemplateName"];
    [EMClient.sharedClient.pushManager setPushTemplate:pushTemplateName completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)getPushTemplate:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPushTemplate:^(NSString * _Nullable aPushTemplateName, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aPushTemplateName];
    }];
}


#pragma mark - 481
- (void)syncConversationsSilentMode:(NSDictionary *)param
                        channelName:(NSString *)aChannelName
                             result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager syncSilentModeConversationsFromServerCompletion:^(EMError * _Nullable error) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:error
                           object:nil];
    }];
}

- (void)bindDeviceToken:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *notifierName = param[@"notifierName"];
    EMClient.sharedClient.options.apnsCertName = notifierName;
    NSString *deviceToken = param[@"deviceToken"];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [EMClient.sharedClient bindFCMToken:deviceToken completion:^(EMError * _Nullable aError) {
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:nil];
        }];
    });
}

@end
